docs/src/03-tutorials/03-pregel-tutorial.md

```diff
@@ -0,0 +1,617 @@
+vertices_simple = spark.createDataFrame([
+    ("A", "Alice"),
+    ("B", "Bob"),
+    ("C", "Charlie"),
+    ("D", "David"),
+], ["id", "name"])
+
+edges_simple = spark.createDataFrame([
+    ("A", "B", "follows"),
+    ("A", "C", "follows"),
+    ("B", "C", "follows"),
+    ("C", "D", "follows"),
+], ["src", "dst", "relationship"])
+
+# Add initial degree column
+vertices_simple = vertices_simple.withColumn("start_degree", F.lit(1))
+g_simple = GraphFrame(vertices_simple, edges_simple)
+
+# Calculate in-degree using AggregateMessages
+msgToDst = AM.src["start_degree"]
+in_degrees = g_simple.aggregateMessages(
+    F.sum(AM.msg).alias("in_degree"),
+    sendToDst=msgToDst)
+
+# Join with all vertices and fill missing values with 0
+complete_degrees = (
+    g_simple.vertices
+    .join(in_degrees, on="id", how="left")
+    .na.fill(0, ["in_degree"])
+    .select("id", "name", "in_degree")
+)
+
+complete_degrees.orderBy("id").show()
+
+# Expected output:
+# +---+-------+---------+
+# | id|   name|in_degree|
+# +---+-------+---------+
+# |  A|  Alice|        0|  (no incoming edges)
+# |  B|    Bob|        1|  (A->B)
+# |  C|Charlie|        2|  (A->C, B->C)
+# |  D|  David|        1|  (C->D)
+# +---+-------+---------+
+```
+
+This simple example clearly shows how AggregateMessages works:
+- Node A has 0 in-degree (no one follows Alice)
+- Node B has 1 in-degree (Alice follows Bob)
+- Node C has 2 in-degree (Alice and Bob follow Charlie)
+- Node D has 1 in-degree (Charlie follows David)
+
+## Introducing Pregel: In-Degree Calculation
+
+Now let's implement the **same** in-degree calculation using Pregel. This helps us understand Pregel's API by comparing it with AggregateMessages:
+
+```python
+from graphframes.lib import Pregel
+
+# Using the same simple test graph
+vertices_simple = spark.createDataFrame([
+    ("A", "Alice"),
+    ("B", "Bob"),
+    ("C", "Charlie"),
+    ("D", "David"),
+], ["id", "name"])
+
+edges_simple = spark.createDataFrame([
+    ("A", "B", "follows"),
+    ("A", "C", "follows"),
+    ("B", "C", "follows"),
+    ("C", "D", "follows"),
+], ["src", "dst", "relationship"])
+
+g_simple = GraphFrame(vertices_simple, edges_simple)
+
+# Calculate in-degree using Pregel API
+pregel_result = g_simple.pregel \
+    .setMaxIter(1) \
+    .withVertexColumn(
+        "in_degree",                      # Column name
+        F.lit(0),                         # Initial value: start with 0
+        F.coalesce(Pregel.msg(), F.lit(0))  # Update: use received message or keep 0
+    ) \
+    .sendMsgToDst(F.lit(1)) \
+    .aggMsgs(F.sum(Pregel.msg())) \
+    .run()
+
+pregel_result.select("id", "name", "in_degree").orderBy("id").show()
+
+# Output:
+# +---+-------+---------+
+# | id|   name|in_degree|
+# +---+-------+---------+
+# |  A|  Alice|        0|
+# |  B|    Bob|        1|
+# |  C|Charlie|        2|
+# |  D|  David|        1|
+# +---+-------+---------+
+```
+
+### Understanding the Pregel API
```

Comment:
I was expecting you will highlight the whole API... Early stopping conditions especially. For example, PageRank may be run for a fixed num of iterations, but also until convergence based on tolerance factor. I was thinking you put it to the end, but you didn't. Can you add a couple of words about different early stopping strategies and maybe a simple example?